### PR TITLE
refactor: using match instead of if-else

### PR DESCRIPTION
**Pull request by ShoeheyOt**
Change the control of flow from if-else to match for high robustness.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -15,72 +15,75 @@ fn main() -> io::Result<()> {
...
-            let mut new_todo = String::new();
-            println!("please input new todo >> \n");
-            io::stdout().flush()?;
-            io::stdin().read_line(&mut new_todo)?;
+                println!("\n");
 
-            match try_add_lines(&new_todo) {
-                Err(why) => eprintln!("couldn't add {} : {}", new_todo, why),
-                Ok(_) => println!("success!\n"),
+                continue;
```
Comment (hitohata) [inline src/main.rs]: Unnecessary.

File: src/main.rs [hunk 2]
```diff
@@ -15,72 +15,75 @@ fn main() -> io::Result<()> {
...
-                    for todo in vector {
-                        println!("{}", todo)
-                    }
-                } else {
-                    eprintln!("couldn't read file!!!")
+                match try_add_lines(&new_todo) {
+                    Err(why) => eprintln!("couldn't add {} : {}", new_todo, why),
+                    Ok(_) => println!("success!\n"),
                 }
+                continue;
```
Comment (hitohata) [inline src/main.rs]: Unnecessary.

File: src/main.rs [hunk 3]
```diff
@@ -15,72 +15,75 @@ fn main() -> io::Result<()> {
...
-            break;
+                continue;
+            }
+            "4" => {
+                println!("See you!");
+                break;
+            }
+            _ => {
+                println!("type again\n");
+                continue;
```
Comment (hitohata) [inline src/main.rs]: Unnecessary.

File: src/main.rs [hunk 4]
```diff
@@ -15,72 +15,75 @@ fn main() -> io::Result<()> {
...
                 }
-            }
-
-            continue;
-        }
 
-        if input.trim() == "4" {
-            println!("See you!");
-            break;
+                continue;
```
Comment (hitohata) [inline src/main.rs]: Unnecessary.

File: src/main.rs [hunk 5]
```diff
@@ -15,72 +15,75 @@ fn main() -> io::Result<()> {
...
+                } else {
+                    println!("which one? >");
 
-                let mut delete_todo = String::new();
+                    if let Ok(vector) = try_read_lines("todo.txt") {
+                        for todo in vector {
+                            println!("{}", todo)
+                        }
+                    } else {
+                        eprintln!("couldn't read file!!!")
```
Comment (hitohata) [inline src/main.rs]: In this case, it shouldn't go through the following process. Am I correct?
```suggestion
                        eprintln!("couldn't read file!!!")
                        continue;
```

File: src/main.rs [hunk 6]
```diff
@@ -15,72 +15,75 @@ fn main() -> io::Result<()> {
...
+                    Err(why) => eprintln!("couldn't add {} : {}", new_todo, why),
+                    Ok(_) => println!("success!\n"),
                 }
+                continue;
+            }
+            "3" => {
+                if try_read_lines("todo.txt")?.len() == 0 {
+                    println!("Nothing to delete now \n");
+                    io::stdout().flush()?;
+                } else {
```
Comment (hitohata) [inline src/main.rs]:
```suggestion
               let todos = match try_read_lines("todo.txt") {
                    Ok(lines) => lines,
                    Err(_) => {
                        eprintln!("couldn't read file!!!");
                        continue;
                    }
                };

                if todos.len() == 0 {
                    println!("Nothing to delete now \n");
                    io::stdout().flush()?;
                    continue
                }

                println!("which one? >");

                for todo in todos {
                    println!("{}", todo)
                }
```

File: src/main.rs
Comment (hitohata) [inline src/main.rs:1]: Do you need \n in line 12?
I think it isn't needed. This is my preference though.
```rs
print!("Please type number -> ");
```

**Review by hitohata — COMMENTED** [review]
Please check comments?

File: src/main.rs (hunk 6 above)
Comment (ShoeheyOt) [reply src/main.rs]: ~~Could you provide me what to do with this part?
also I think this code is old version... I don't know where you get them from...~~
a, now I understand it , please ignore above comment, sorry

File: src/main.rs (hunk 1 above)
Comment (ShoeheyOt) [reply src/main.rs]: thanks! Ideleted it.

File: src/main.rs (hunk 2 above)
Comment (ShoeheyOt) [reply src/main.rs]: Deleted!

File: src/main.rs (hunk 3 above)
Comment (ShoeheyOt) [reply src/main.rs]: Deleted!

File: src/main.rs (hunk 4 above)
Comment (ShoeheyOt) [reply src/main.rs]: Deleted!

File: src/main.rs (hunk 5 above)
Comment (ShoeheyOt) [reply src/main.rs]: Yes, you are right, I changed the way to store Vector<String> to variable in order to avoide duplicate error handling

File: src/main.rs
Comment (ShoeheyOt) [reply src/main.rs:1]: No, in this message I don't think it is needed to add a new line.

**Comment by ShoeheyOt** [comment]
I appreciate your review. I updated what you pointed out. Please review when you have time.

File: src/file_operation.rs [hunk 7]
```diff
@@ -87,13 +88,13 @@ pub fn try_add_lines(new_todo: &str) -> io::Result<()> {
 ///     None => println!("Todo not found or error reading file"),
 ///  }
 /// ```
-pub fn find_index_opt(searched_todo: &String) -> Option<usize> {
+pub fn find_index_opt(searched_todo: &String) -> Result<Option<usize>, Error> {
     match try_read_lines("todo.txt") {
         Err(why) => {
-            eprintln!("couldn't read the lines : {}", why);
-            None
+            println!("couldn't read the lines : {}", why);
```
Comment (hitohata) [inline src/file_operation.rs]: Why do you print here? You can let a caller handle how to treat this error.

File: src/main.rs [hunk 8]
```diff
@@ -15,72 +15,73 @@ fn main() -> io::Result<()> {
...
+                };
 
                 let mut delete_todo = String::new();
 
                 io::stdin().read_line(&mut delete_todo)?;
 
                 match find_index_opt(&delete_todo) {
-                    Some(_) => {
+                    Ok(_) => {
                         try_update_file(&delete_todo)?;
```
Comment (hitohata) [inline src/main.rs]: If you return `Optional`, I think you need to handle the case that the function returns None.
In this case, when the function returns None, this flow goes to print successfully~.

File: src/main.rs
Comment (hitohata) [reply src/main.rs:1]: This commend has not been fixed.

**Review by hitohata — COMMENTED** [review]
I think you misunderstand the `print` and the `println`. The `println` puts a new line after printing the str in the argument. 
Maybe you don't need to add `\n` in some points.
https://doc.rust-lang.org/rust-by-example/hello/print.html

File: src/main.rs
Comment (ShoeheyOt) [reply src/main.rs:1]: Thank you for explanation, now I understand and update it

File: src/file_operation.rs (hunk 7 above)
Comment (ShoeheyOt) [reply src/file_operation.rs]: You are right, it does not need to print here.

File: src/main.rs (hunk 8 above)
Comment (ShoeheyOt) [reply src/main.rs]: You mean it is possible that the function returns `OK` but `None`, right? If so, I can change the code to either return `Result<usize, Error>` in `find_index_opt()` or handle what to do if return `None`,  I think the latter is better in point of detailed error handling. I would appreciate if you give me an advice.

Comment (hitohata) [reply src/main.rs]: You should write the latter way.

Comment (hitohata) [reply src/main.rs]: The error will occur when reading file process is failed. The None will be returned the input word is not found.

File: src/main.rs [hunk 9]
```diff
@@ -9,78 +9,79 @@ fn main() -> io::Result<()> {
...
                         println!("successfully delete {}", delete_todo);
                     }
-                    None => println!("not found\n"),
+                    Err(why) => eprintln!("not found : {}", why),
```
Comment (hitohata) [inline src/main.rs]: I feel this comment is not appropriate. In this case, the Err is returned when the file, todo.txt, is not found.

File: src/main.rs (hunk 8 above)
Comment (hitohata) [reply src/main.rs]: You put `?` many points. Do you understand what it means? You might misunderstand the difference between try-catch and result or optional.
https://doc.rust-lang.org/rust-by-example/std/result/question_mark.html

File: src/main.rs (hunk 9 above)
Comment (ShoeheyOt) [reply src/main.rs]: Yes, I updated the text to tell it failed to read the file

File: src/main.rs (hunk 8 above)
Comment (ShoeheyOt) [reply src/main.rs]: My understanding for `?`  is shorthand of `match` expression for concise code. try-catch is if an error happens in try block, jump into catch block and not execution in try block. Result is explicity implement what to do when error occurs, Optional is return value => `Some`, no value => `None`

Comment (hitohata) [reply src/main.rs]: OK, you understand well. By the way, won't you handle `try_update_file`?

Comment (ShoeheyOt) [reply src/main.rs]: You mean handling Error and Ok explicitliy, like below? If so, yes, I think I should do so.
From this,
```Rust
src/main.rs
match find_index_opt(&delete_todo) {
                    Ok(index) => match index {
                        None => println!("not found in the list"),

                        Some(_) => {
                            try_update_file(&delete_todo)?;
                            println!("successfully delete {} ", delete_todo);
                        }
                    },
                    Err(why) => eprintln!("couldn't read the file : {}", why),
                }

```

to this
```Rust
 match find_index_opt(&delete_todo) {
                    Ok(index) => match index {
                        None => println!("not found in the list"),

                        Some(_) => match try_update_file(&delete_todo) {
                            Err(why) => eprintln!("failed to update the file : {}", why),
                            Ok(_) => println!("successfully delete {}", delete_todo),
                        },
                    },
                    Err(why) => eprintln!("couldn't read the file : {}", why),
                }

```

Comment (ShoeheyOt) [reply src/main.rs]: Thank you for confirming about thi verbally today, I already updated it.

**Comment by ShoeheyOt** [comment]
Thank you for reviewing this PR, I will merge for this one as discussed.

**Pull request merged**